ee/modules/025-static-routing-manager/images/agent/pkg/controller/routes_reconciler_agent.go

```diff
@@ -222,44 +222,171 @@ func periodicalRunEventReconcile(
+		preparedNlRoute.Gw = net.ParseIP(re.gateway)
+	}
+	if re.dev != "" && re.devId != 0 {
+		preparedNlRoute.LinkIndex = re.devId
+	}
+
+	return preparedNlRoute, nil
+}
+
+func getRouteEntryFromNetlinkRoute(nlRoute netlink.Route) (RouteEntry, error) {
+	preparedRE := RouteEntry{}
+
+	if nlRoute.Dst != nil {
+		preparedRE.destination = nlRoute.Dst.String()
+	}
+	if nlRoute.Gw != nil {
+		preparedRE.gateway = nlRoute.Gw.String()
+	}
+	if nlRoute.LinkIndex > 0 {
+		preparedRE.devId = nlRoute.LinkIndex
+		link, err := netlink.LinkByIndex(nlRoute.LinkIndex)
+		if err != nil {
+			return RouteEntry{}, fmt.Errorf("can not find Link by Index %v, err: %w",
+				nlRoute.LinkIndex,
+				err,
+			)
+		}
+		preparedRE.dev = link.Attrs().Name
+	}
+	if nlRoute.Table > 0 {
+		preparedRE.table = nlRoute.Table
+	}
+
+	return preparedRE, nil
+}
+
+func getRouteEntryFromRouteAndTable(route v1alpha1.Route, tbl int) (RouteEntry, error) {
+	preparedRE := RouteEntry{
+		destination: route.Destination,
+		table:       tbl,
+	}
+
+	if route.Gateway != "" {
+		preparedRE.gateway = route.Gateway
+	}
+	if route.Dev != "" {
+		link, err := netlink.LinkByName(route.Dev)
+		if err != nil {
+			return RouteEntry{}, fmt.Errorf("can not find Link by Name %v, err: %w",
+				route.Dev,
+				err,
+			)
+		}
+		preparedRE.dev = route.Dev
+		preparedRE.devId = link.Attrs().Index
 	}
+	if route.Gateway != "" && route.Dev == "" {
+		// gwrt, err := netlink.RouteGetWithOptions(net.ParseIP(re.gateway), nil)
+		gwRts, err := netlink.RouteGet(net.ParseIP(preparedRE.gateway))
+		if err != nil || len(gwRts) == 0 {
+			return RouteEntry{}, fmt.Errorf("can not find egress Link by route to Gateway %v, err: %w",
+				route.Gateway,
+				err,
+			)
+		}
+		if len(gwRts) > 1 {
+			fmt.Errorf("more then one egress Link find for Gateway %v. Used only first: %v",
+				route.Gateway,
+				gwRts[0].String(),
+			)
+		}
+		preparedRE.devId = gwRts[0].LinkIndex
 
-	return fmt.Sprintf("%v", hash)
+		link, err := netlink.LinkByIndex(gwRts[0].LinkIndex)
+		if err == nil {
+			preparedRE.dev = link.Attrs().Name
+		} else {
+			return RouteEntry{}, fmt.Errorf("can not find Link by Index %v (for Gateway %v), err: %w",
+				gwRts[0].LinkIndex,
+				route.Gateway,
+				err,
+			)
+		}
+	}
+	return preparedRE, nil
 }
 
 // RouteEntryMap: type, service functions and methods
 
 type RouteEntryMap map[string]RouteEntry
 
-func (rem *RouteEntryMap) AppendRE(re RouteEntry) {
-	if len(*rem) == 0 {
-		*rem = make(map[string]RouteEntry)
+func (rem *RouteEntryMap) getRoutes() v1alpha1.Routes {
+	preparedRoutes := make([]v1alpha1.Route, 0)
+	for _, re := range *rem {
+		re.getRoute()
+		preparedRoutes = append(preparedRoutes, re.getRoute())
+	}
+	return v1alpha1.Routes{
+		Routes: preparedRoutes,
 	}
-	(*rem)[re.getHash()] = re
 }
 
-func (rem *RouteEntryMap) AppendR(route v1alpha1.Route, tbl int) {
+func (rem *RouteEntryMap) AppendRE(re RouteEntry) {
 	if len(*rem) == 0 {
 		*rem = make(map[string]RouteEntry)
 	}
```

Comment:
```suggestion
```

Reply:
Better to use:
```
func NewIPRuleEntryMap() *IPRuleEntryMap {
	ipRuleEntryMap := make(IPRuleEntryMap)
	return &ipRuleEntryMap
}
g := NewIPRuleEntryMap()
(*g)["something"] = IPRuleEntry{...}
// Use explicit call NewIPRuleEntryMap for initialize in all place
```

Reply:
Or sync.Once instead `if len(*rem) == 0` but  it doesn't look good too.

Reply:
fixed